index.ts

```diff
@@ -14,23 +15,35 @@ abstract class NgComponent<Props, State> {
     }
   */
   // nb: this method is explicity exposed for unit testing
-  public $onChanges(changes: object) {
+  $onChanges(changes: object) {
     const oldProps = mapValues<{}, Props>(changes, 'previousValue')
     const newProps = mapValues<{}, Props>(changes, 'currentValue')
+    if (!this.__isFirstRender && !this.didPropsChange(newProps, oldProps)) return
+    this.lifecycle(newProps, {} as State)
+  }
 
+  // note: this isn't meant to exactly replicate React's setState (e.g. http://stackoverflow.com/q/28922275/7816712)
+  protected setState(newState: State) {
+    if (this.__setStateTriggersLifecycle) {
+      setTimeout(() => this.lifecycle({} as Props, newState), 0)
+    } else {
+      this.state = assign({}, this.state, newState)
+    }
+  }
+
+  private lifecycle(newProps: Props, newState: State) {
     const nextProps = assign({}, this.props, newProps)
-    // TODO: implement nextState (which also means implement this.setState)
+    const nextState = assign({}, this.state, newState)
 
     if (this.__isFirstRender) {
-      assign(this, { props: nextProps })
-      this.componentWillMount()
+      assign(this, { props: nextProps, state: nextState })
+      this.avoidSetStateLifecycle(() => this.componentWillMount())
       this.render()
       this.__isFirstRender = false
     } else {
-      if (!this.didPropsChange(newProps, oldProps)) return
-      this.componentWillReceiveProps(nextProps)
-      const shouldUpdate = this.shouldComponentUpdate(nextProps, this.state)
-      assign(this, { props: nextProps })
+      this.avoidSetStateLifecycle(() => this.componentWillReceiveProps(nextProps))
```

Comment:
Can you add a test that `this.state` still refers to the old state inside componentWillReceiveProps